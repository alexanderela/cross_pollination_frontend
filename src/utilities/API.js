export const fetchData = async (url) => {
  const response = await fetch(url);
  const data = await response.json();
  return data;
  //changed name from jsonData to data. The data is no longer json at this point .json() will parse it. It is usable, just wrapped in a promise.
};

export const addUser = async (name, email, password) => {
  try {
    const url = 'https://flagz4u.herokuapp.com/signup'
    const response = await fetch(url, {
      method: 'POST',
      body: JSON.stringify({name: name, email: email, password: password}),
      headers: {
        "Content-Type": "application/json"
      }
    })
    const result = await response.json();
    const user = {id: result.id}
    return user
  } catch(error) {
    throw new Error()
  }
}

export const getUser = async (email, password) => {
  const url = 'https://flagz4u.herokuapp.com/sigin'
  const response = await fetchData(url);
  const matchingUser = response.data.find(user => user.email === email && user.password === password)
  if(matchingUser) {
    const user = {...matchingUser}
    return user
  } else {
    throw new Error()
  }
} 
