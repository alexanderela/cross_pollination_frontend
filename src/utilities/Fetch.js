import * as API from './API'
import { buildQuestion, checkCountry } from './helper'

export const fetchCorrectCountry = async (randomId, usedCountries) => {
  const url = `https://world-of-flags-backend.herokuapp.com/api/v1/country/${randomId}`
  const countryArray = await API.fetchData(url)
  const correctCountry = countryArray[0]
  const countryAlreadyExists = checkCountry(correctCountry.name, usedCountries)

  if (countryAlreadyExists) {
    let randomNumber = Math.floor(Math.random() * (196 - 1) + 1)
    return fetchCorrectCountry(randomNumber, usedCountries)
  } else {
    const countryFacts = await fetchCountryFacts(randomId)
    return buildQuestion(correctCountry, countryFacts)
  }
}

export const fetchCountryFacts = async countryId => {
  const url = `https://world-of-flags-backend.herokuapp.com/api/v1/facts/${countryId}`

  const countryFacts = await API.fetchData(url)
  return countryFacts
}
