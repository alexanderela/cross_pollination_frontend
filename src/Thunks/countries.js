import { setCurrentCountry } from '../actions/countryActions'
import { updateUsedCountries } from '../actions/usedCountryActions'
import { hasErrored } from '../actions/hasErroredAction'
import * as Fetch from '../utilities/Fetch'

export const getCurrentCountry = (randomId, usedCountries) => {
  return async dispatch => {
    try {
      const currentCountry = await Fetch.fetchCorrectCountry(
        randomId,
        usedCountries
      );
      dispatch(setCurrentCountry(currentCountry));
      dispatch(updateUsedCountries(currentCountry.name));
    } catch (error) {
      dispatch(hasErrored(true))
    }
  };
};
