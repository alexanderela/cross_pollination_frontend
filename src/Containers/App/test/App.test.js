import React from 'react'
import ReactDOM from 'react-dom'
import { App, mapStateToProps, mapDispatchToProps } from '../index'
import { shallow, mount } from 'enzyme'
import * as Fetch from '../../../utilities/Fetch'
import { getCurrentCountry } from '../../../Thunks/countries.js';

describe('App', () => {
  let wrapper
  let mockNewPoints
  let mockUsedCountries
  let mockCountry

  beforeEach(() => {
    mockNewPoints = 3

    mockUsedCountries = ['Mexico', 'Hungary', 'Ireland', 'Sweden']

    mockCountry = {
      "id": 79,
      "name": "Turkmenistan",
      "flag": "/images/flags/turkmenistan.png",
      "country_outline": "/images/outlines/turkmenistan.png",
      "created_at": "2018-12-23T16:48:16.005Z",
      "updated_at": "2018-12-23T16:48:16.005Z",
      "multipleChoice": [
        "France",
        "Nigeria",
        "Turkmenistan",
        "Japan"
      ]
    };

    wrapper = shallow(<App 
                        currentCountry={mockCountry}
                        user={{id: 1, loggedIn: true}}
                        usedCountries={mockUsedCountries}
                        getCurrentCountry = {jest.fn().mockImplementation(() => {})}
                      />)
  });

  it.skip('should render like the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  describe('getCountry', () => {
    it('should call getCurrentCountry with', async () => {
      await wrapper.instance().getCountry()

      expect(wrapper.instance().props.getCurrentCountry).toHaveBeenCalled()
    })
  })

  describe('mapStateToProps', () => {
    let mockState = {
      user: {id: 2, user: 'Bob', email: "bob@bob.com", loggedIn: true},
      currentCountry: {
        id: 79,
        name: 'Turkmenistan',
        flag: '/images/flags/turkmenistan.png',
        country_outline: '/images/outlines/turkmenistan.png',
        multipleChoice: ['France', 'Nigeria', 'Turkmenistan', 'Japan'],
      },
      usedCountries: ['Mexico', 'Hungary', 'Ireland', 'Sweden'],
    }

    it('should return a user in the props object', () => {
      const expected = {id: 2, user: 'Bob', email: "bob@bob.com", loggedIn: true}
        
      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps.user).toEqual(expected)
    })

    it('should return a currentCountry in the props object', () => {
      const expected = {
        id: 79,
        name: 'Turkmenistan',
        flag: '/images/flags/turkmenistan.png',
        country_outline: '/images/outlines/turkmenistan.png',
        multipleChoice: ['France', 'Nigeria', 'Turkmenistan', 'Japan'],
      }

      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps.currentCountry).toEqual(expected)
    })
    it('should return a usedCountries array in the props object', () => {
      const expected = ['Mexico', 'Hungary', 'Ireland', 'Sweden']

      const mappedProps = mapStateToProps(mockState)
      expect(mappedProps.usedCountries).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    let mockDispatch;

    beforeEach(() => {
      mockDispatch = jest.fn()
    })

    it('should map a key of getCurrentCountry', () => {
      const dispatchProps = mapDispatchToProps(mockDispatch);
      expect(dispatchProps.getCurrentCountry).toBeDefined();
    })

    it('should have getCurrentCountry call dispatch', () => {
      const dispatchProps = mapDispatchToProps(mockDispatch);
      dispatchProps.getCurrentCountry(79, mockUsedCountries);
      expect(mockDispatch).toHaveBeenCalled();
    })
  });
});