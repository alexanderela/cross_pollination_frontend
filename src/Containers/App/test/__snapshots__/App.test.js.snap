// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render like the snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App
    currentCountry={
      Object {
        "country_outline": "/images/outlines/turkmenistan.png",
        "created_at": "2018-12-23T16:48:16.005Z",
        "flag": "/images/flags/turkmenistan.png",
        "id": 79,
        "multipleChoice": Array [
          "France",
          "Nigeria",
          "Turkmenistan",
          "Japan",
        ],
        "name": "Turkmenistan",
        "updated_at": "2018-12-23T16:48:16.005Z",
      }
    }
    setCurrentCountry={[MockFunction]}
    updateUsedCountries={[MockFunction]}
    usedCountries={
      Array [
        "Mexico",
        "Hungary",
        "Ireland",
        "Sweden",
      ]
    }
    user={
      Object {
        "id": 1,
        "loggedIn": true,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Switch>
        <Route
          exact={true}
          path="/"
          render={[Function]}
        />
        <Route
          exact={true}
          path="/account"
          render={[Function]}
        />
      </Switch>,
      "className": "App",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Route
            exact={true}
            path="/"
            render={[Function]}
          />,
          <Route
            exact={true}
            path="/account"
            render={[Function]}
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "exact": true,
            "path": "/account",
            "render": [Function],
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Switch>
          <Route
            exact={true}
            path="/"
            render={[Function]}
          />
          <Route
            exact={true}
            path="/account"
            render={[Function]}
          />
        </Switch>,
        "className": "App",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Route
              exact={true}
              path="/"
              render={[Function]}
            />,
            <Route
              exact={true}
              path="/account"
              render={[Function]}
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "exact": true,
              "path": "/",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "exact": true,
              "path": "/account",
              "render": [Function],
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
